### PR TITLE
feat: I can interrupt an expression evaluation

### LINKED ISSUE #823: I can't interrupt an expression evaluation
**Is your feature request related to a problem? Please describe.**
The feel engine is currently stuck while executing long-running expressions because of the lack of "cooperation mode" ([ref](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html)). It is not possible to interrupt the execution from outside the engine. 

Related forum discussion: https://forum.camunda.io/t/feelengines-cooperation-mode-while-performing-longrunning-and-resource-consuming-operations/51049

> We are using org.camunda.feel.FeelEngine instance to evaluate FEEL expressions and trying to find a proper way to “sandbox” this operation and give a dedicated timeout for each one, in case the user tries to execute a malicious expression.
We’ve came across the following expression as an example of such harmful snippet which can bring the instance down:
>
> count(for x in 1..(2 ** 16) return {"power": 2 ** x}) > 0

**Describe the solution you'd like**
The most classical solution is to repeatedly check whether the current thread was interrupted and throw an appropriate exception

### PR DESCRIPTION
**Pull request by vicmosin**
## Description

This simple change allows long-running feel evaluation/parsing processes to be gracefully interrupted. Consumers can execute the FEEL engine parsing/evaluations on a separate thread and be able to cancel them interruptedly.

## Related issues

closes #823

### CODE REVIEW

**Comment by vicmosin** [comment]
The solution allows consumers to execute the feel-engine parsing/evaluations on a separate threads and be able to cancel them itnerruptedly

File: src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala
Comment (saig0) [inline src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala:45]: :wrench: We should add a comment to explain why we are checking for the interruption. 

It adds context for the next devs who is looking at this code.

File: src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala
Comment (saig0) [inline src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala:1]: :x: We should add a test case to verify that the interruption is working. And, to make sure that is will not be removed accidentally in the future.

**Review by saig0 — CHANGES_REQUESTED** [review]
@vicmosin your changes look good. :tada: I added a few comments we must address before merging. Please have a look.

File: src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala
Comment (vicmosin) [reply src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala:1]: @saig0 Done, but not sure it's the proper implementation, I am not that familiar with scala

**Comment by vicmosin** [comment]
Hey @saig0 , once merged can we get it as part of next minor release, let's say 1.17.6 ?

**Comment by vicmosin** [comment]
@saig0 anything else I should adjust?

File: src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala:34]: :wrench: `EvaluationResultMatchers` is not used. We can remove it.

```suggestion
class InterpreterInterruptionTest extends AnyFlatSpec with Matchers with FeelEngineTest {
```

File: src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala:55]: :wrench: We could decrease the waiting time to speed up the test. We don't need to wait so long. 

```suggestion
    Thread.sleep(100) // Let evaluation start
```

File: src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala:57]: :wrench: We could use the return value of `await()` to verify that the thread was interrupted. 

```suggestion
    val wasInterrupted = countDownLatch.await(1, TimeUnit.SECONDS)

    wasInterrupted should be(true)
```

File: src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala
Comment (saig0) [inline src/test/scala/org/camunda/feel/impl/interpreter/InterpreterInterruptionTest.scala:59]: :wrench: Using a thread dump to verify the interruption is a nice idea. However, it is a bit too much for this test case. It is enough to verify that the thread was interrupted by checking the count-down-latch.

Let's remove this part to keep it easy.

**Review by saig0 — CHANGES_REQUESTED** [review]
@vicmosin very nice. :rocket: I see some improvements for the test case. Please have a look. :cookie:

**Comment by vicmosin** [comment]
@saig0 done

**Review by saig0 — APPROVED** [review]
@vicmosin great. :zap: LGTM. :tada:

**Comment by vicmosin** [comment]
@saig0 when to wait for a release, approx obviously?

**Comment by saig0** [comment]
Successfully created backport PR for `1.17`:
- #830

**Comment by saig0** [comment]
@vicmosin I will build a patch release for 1.17 tomorrow. :building_construction:

**Pull request merged**
